app/controllers/trainees/confirm_details_controller.rb

```diff
@@ -12,24 +12,31 @@ def show
 
       if trainee.draft?
         @confirm_detail_form = ConfirmDetailForm.new(mark_as_completed: trainee.progress.public_send(trainee_section_key))
-        @missing_data_view = MissingDataView.new(form_klass.new(trainee))
       end
 
+      @review_summary = ReviewSummary::View.new(form: form, trainee: trainee, data_view: MissingDataView.new(form_klass.new(trainee)))
       @confirmation_component = component_klass.new(data_model: trainee.draft? ? trainee : form_klass.new(trainee))
     end
 
     def update
-      form_klass.new(trainee).save! unless trainee.draft?
-
-      toggle_trainee_progress_field if trainee.draft?
-
-      flash[:success] = "Trainee #{flash_message_title} updated"
-
-      redirect_to page_tracker.last_origin_page_path || trainee_path(trainee)
+      if form.valid?
+        form_klass.new(trainee).save! unless trainee.draft?
```

Comment:
You've created a new method called `form`. You can take advantage of this:

```suggestion
        form.save! unless trainee.draft?
```